crates/ironrdp-dvc/src/server.rs

```diff
@@ -0,0 +1,255 @@
+                if c.state != ChannelState::Creation {
+                    return Err(invalid_message_err!("DRDYNVC", "", "invalid channel state"));
+                }
+                if create_resp.creation_status != dvc::DVC_CREATION_STATUS_OK {
+                    c.state = ChannelState::CreationFailed(create_resp.creation_status);
+                    return Ok(resp);
+                }
+                c.state = ChannelState::Opened;
+                let msg = c.processor.start(create_resp.channel_id)?;
+                resp.extend(encode_dvc_data(id, msg)?);
+            }
+            dvc::ClientPdu::CloseResponse(close_resp) => {
+                debug!("Got DVC Close Response PDU: {close_resp:?}");
+                let c = self.channel_by_id(close_resp.channel_id)?;
+                if c.state != ChannelState::Opened {
+                    return Err(invalid_message_err!("DRDYNVC", "", "invalid channel state"));
+                }
+                c.state = ChannelState::Closed;
+            }
+            dvc::ClientPdu::DataFirst(data) => {
+                let c = self.channel_by_id(data.channel_id)?;
+                if c.state != ChannelState::Opened {
+                    return Err(invalid_message_err!("DRDYNVC", "", "invalid channel state"));
+                }
+                if let Some(complete) = c
+                    .data
+                    .process_data_first_pdu(data.total_data_size as usize, dvc_ctx.dvc_data.into())
+                {
+                    let msg = c.processor.process(data.channel_id, &complete)?;
+                    resp.extend(encode_dvc_data(data.channel_id, msg)?);
+                }
+            }
+            dvc::ClientPdu::Data(data) => {
+                let c = self.channel_by_id(data.channel_id)?;
+                if c.state != ChannelState::Opened {
+                    return Err(invalid_message_err!("DRDYNVC", "", "invalid channel state"));
+                }
+                if let Some(complete) = c.data.process_data_pdu(dvc_ctx.dvc_data.into()) {
+                    let msg = c.processor.process(data.channel_id, &complete)?;
+                    resp.extend(encode_dvc_data(data.channel_id, msg)?);
+                }
+            }
+        }
+
+        Ok(resp)
+    }
+}
+
+impl SvcServerProcessor for DrdynvcServer {}
+
+struct DynamicChannelCtx<'a> {
+    dvc_pdu: vc::dvc::ClientPdu,
+    dvc_data: &'a [u8],
+}
+
+fn decode_dvc_message(user_data: &[u8]) -> PduResult<DynamicChannelCtx<'_>> {
+    let mut user_data = user_data;
+    let user_data_len = user_data.len();
+
+    // … | dvc::ClientPdu | …
+    let dvc_pdu =
+        vc::dvc::ClientPdu::from_buffer(&mut user_data, user_data_len).map_err(|e| custom_err!("DVC client PDU", e))?;
+
+    // … | DvcData ]
+    let dvc_data = user_data;
+
+    Ok(DynamicChannelCtx { dvc_pdu, dvc_data })
+}
+
+fn encode_dvc_message(pdu: vc::dvc::ServerPdu) -> PduResult<SvcMessage> {
+    // FIXME: use PduEncode instead
```

Comment:
Should this `FIXME` be removed or fixed?

Reply:
I am working on it, but it's another branch.. tedious conversion from legacy to/from buffer to encode/decode trait...